File: app/src/main/java/com/example/rcapp/LeDeviceListAdapter.java
package com.example.rcapp;

import static android.Manifest.permission.BLUETOOTH_CONNECT;

import android.content.Context;
import android.content.pm.PackageManager;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.app.ActivityCompat;

import java.util.ArrayList;
import java.util.List;
import android.bluetooth.BluetoothDevice;

public class LeDeviceListAdapter extends ArrayAdapter<BluetoothDevice> {
    private final List<BluetoothDevice> bluetoothList;
    private final Context context;

    public LeDeviceListAdapter(@NonNull Context context, int resource) {
        super(context, resource);
        this.context = context;
        this.bluetoothList = new ArrayList<>();
    }

    public void addDevice(BluetoothDevice device) {
        if (!bluetoothList.contains(device)) {
            bluetoothList.add(device);
            notifyDataSetChanged();
        }
    }

    public void clearList() {
//        bluetoothList.remove(1);
        bluetoothList.clear();
        Log.e("LifeCycle", "scanning1.5");

        notifyDataSetChanged();
        Log.e("LifeCycle", "scanning1.7");

    }

    @Override
    public int getCount() {
        return bluetoothList.size();
    }

    @Override
    public BluetoothDevice getItem(int position) {
        return bluetoothList.get(position);
    }

    @NonNull
    @Override
    public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
        ViewHolder viewHolder;
        if (convertView == null) {
            convertView = LayoutInflater.from(getContext()).inflate(R.layout.bluetooth_item_layout, parent, false);
            viewHolder = new ViewHolder(convertView);
            convertView.setTag(viewHolder);
        } else {
            viewHolder = (ViewHolder) convertView.getTag();
        }

        BluetoothDevice device = getItem(position);
        if (device != null) {
            if (ActivityCompat.checkSelfPermission(context, BLUETOOTH_CONNECT) != PackageManager.PERMISSION_GRANTED) {
                ((BluetoothLinkActivity) context).requestPermission(BLUETOOTH_CONNECT);
            } else {
                viewHolder.bleNameTv.setText(device.getName() != null ? device.getName() : "N/A");
            }

            viewHolder.bleLinkBtn.setOnClickListener(v -> {
                if (context instanceof BluetoothLinkActivity) {
                    ((BluetoothLinkActivity) context).BLEDeviceConnect(position);
                }
            });
        }

        return convertView;
    }

    private static final class ViewHolder {
        final TextView bleNameTv;
        final TextView bleLinkBtn;

        ViewHolder(View view) {
            bleNameTv = view.findViewById(R.id.bluetooth_name_tv);
            bleLinkBtn = view.findViewById(R.id.bluetooth_link_btn);
        }
    }
}

File: app/src/main/java/com/example/rcapp/adapter/LeDeviceListAdapter.kt
package com.example.rcapp.adapter

import android.Manifest.permission
import android.annotation.SuppressLint
import android.content.Context
import android.content.pm.PackageManager
import android.bluetooth.le.ScanResult
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.core.app.ActivityCompat
import androidx.recyclerview.widget.RecyclerView
import com.example.rcapp.activity.BluetoothLinkActivity
import com.example.rcapp.databinding.ScanDeviceItemLayoutBinding

class LeDeviceListAdapter(private val context: Context) :
    RecyclerView.Adapter<LeDeviceListAdapter.ViewHolder>() {

    private val bluetoothList: MutableList<ScanResult> =
        ArrayList()
    private val seenDevices : MutableList<String> =
        ArrayList()

    /**
     * 添加设备到列表中，bluetoothList.size-1为最新的设备下标号
     */
    @SuppressLint("MissingPermission")
    fun addDevice(result: ScanResult) {
        val deviceAddress = result.device.address
        if (!seenDevices.contains(deviceAddress)) {
            bluetoothList.add(result)
            seenDevices.add(deviceAddress)
            notifyItemInserted((bluetoothList.size - 1))
        }
    }
    fun clearList() {
        val size = bluetoothList.size
        bluetoothList.clear()
        seenDevices.clear()
        notifyItemRangeRemoved(0, size)
    }

    override fun getItemCount(): Int {
        return bluetoothList.size
    }

    fun getItem(position: Int): ScanResult {
        return bluetoothList[position]
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        // 使用View Binding替代findViewById
        val inflater = LayoutInflater.from(context)
        val binding = ScanDeviceItemLayoutBinding.inflate(inflater, parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val result = getItem(position)
        result.device
        // 检查BLUETOOTH_CONNECT权限
        if (ActivityCompat.checkSelfPermission(context, permission.BLUETOOTH_CONNECT) != PackageManager.PERMISSION_GRANTED) {
            (context as BluetoothLinkActivity).requestPermission(permission.BLUETOOTH_CONNECT)
            return
        }
        holder.binding.bluetoothNameTv.text=result.device.name
        holder.binding.bluetoothAddressTv.text=result.device.address
        //holder.binding.bluetoothAddressTv.text=result.rssi.toString()+" dBm"
        holder.binding.executePendingBindings()
        holder.binding.root.setOnClickListener{
            if (context is BluetoothLinkActivity) {
                BluetoothLinkActivity.bleDeviceConnect(context, position)
            }
        }
        // 设置按钮点击事件
//        holder.binding.bluetoothLinkBtn.setOnClickListener {
//            if (context is BluetoothLinkActivity) {
//                BluetoothLinkActivity.bleDeviceConnect(context, position)
//            }
//        }
    }

    // ViewHolder继承自父类RecyclerView.ViewHolder
    class ViewHolder// super用于调用父类RecyclerView.ViewHolder的构造函数，将视图传递给它
    // 构造函数，接受一个 Binding参数
    internal constructor(val binding: ScanDeviceItemLayoutBinding) :
        RecyclerView.ViewHolder(binding.root)
}